gralloc_gbm.cpp

```diff
@@ -529,3 +543,54 @@ int gralloc_gbm_bo_lock_ycbcr(buffer_handle_t handle,
 
 	return 0;
 }
+
+/*
+ * Check if target device has KMS.
+ */
+bool is_kms_dev(int fd)
+{
+	auto res = drmModeGetResources(fd);
+	if (!res)
+		return false;
+
+	bool is_kms = res->count_crtcs > 0 && res->count_connectors > 0 &&
+			res->count_encoders > 0;
+
+	drmModeFreeResources(res);
+
+	return is_kms;
+}
+
+/*
+ * Search for a KMS device. Return opened file descriptor on success.
+ */
+int open_first_kms_dev(const char *pattern)
+{
+	glob_t glob_buf;
+	memset(&glob_buf, 0, sizeof(glob_buf));
+	int fd;
+
+	int ret = glob(pattern, 0, NULL, &glob_buf);
+	if (ret) {
+		globfree(&glob_buf);
+		return -EINVAL;
+	}
+
+	for (size_t i = 0; i < glob_buf.gl_pathc; ++i) {
+		fd = open(glob_buf.gl_pathv[i], O_RDWR | O_CLOEXEC);
+		if (fd < 0) {
+			ALOGE("failed to open %s with error %s", glob_buf.gl_pathv[i], strerror(errno));
+			continue;
+		}
+
+		if (is_kms_dev(fd))
+			break;
+
+		close(fd);
+		fd = 0;
```

Comment:
Theoretically `0` is a valid fd (if it weren't usually used for `stdin`), this should probably be `-1` (and the `<= 0` above should be `< 0`).